include/ensmallen_bits/de/de_impl.hpp

```diff
@@ -0,0 +1,125 @@
+#ifndef ENSMALLEN_DE_DE_IMPL_HPP
+#define ENSMALLEN_DE_DE_IMPL_HPP
+
+#include "de.hpp"
+#include <random>
+#include <tuple>
+#include <iostream>
+
+namespace ens {
+
+	inline DE::DE( const size_t populationSize ,
+		   const size_t maxGenerations,
+           const double crossoverRate,
+           const double differentialWeight):
+		populationSize(populationSize),
+		maxGenerations(maxGenerations),
+		crossoverRate(crossoverRate),
+		differentialWeight(differentialWeight)
+		{ /*Nothing to do here*/ }
+
+	//!Optimize the function
+	template<typename DecomposableFunctionType>
+  	inline double DE::Optimize(DecomposableFunctionType& function, arma::mat& iterate)
+  	{
+  		//Population Size must be atleast 3 for DE to work
+  		if (populationSize < 3)
+		{
+		  	throw std::logic_error("CNE::Optimize(): population size should be at least"
+		        " 3!");
+		}
+
+  		//Generate a population based on a Gaussian distribution around the given starting point
+		population = arma::randn(iterate.n_rows, iterate.n_cols, populationSize);
+		for(int i = 0; i < populationSize; i++)
+		{
+			arma::mat slicer = population.slice(i);
+			slicer = iterate+slicer;
+			population.slice(i) = slicer;
+		}
+
+		//Initializw helper variables
+		fitnessValues.set_size(populationSize);
+		for(int it = 0; it < populationSize; it++)
+			fitnessValues[it] = function.Evaluate(population.slice(it));
+		double lastBestFitness = fitnessValues.min();
+		arma::mat bestElement;
+		for(int it = 0; it < populationSize; it++)
```

Comment:
Do you think we could move this into the for loop above, so right after we do the ``Evaluate`` we can check if the fitness improved.